### PR TITLE
Add core BBS routes and utilities, update docs

### LINKED ISSUE #1: Auth middleware
Finish auth middleware.

### PR DESCRIPTION
**Pull request by PythonSmall-Q**

### CODE REVIEW

**Comment by PythonSmall-Q** [comment]
帮我测试一下吧，我不确定有没有问题

**Comment by langningchen** [comment]
这是 AI 改的？

**Comment by langningchen** [comment]
这个新的架构感觉挺不错，但是这种更改的量，难道真的是可以用人来审核的吗

**Comment by PythonSmall-Q** [comment]
> 这个新的架构感觉挺不错，但是这种更改的量，难道真的是可以用人来审核的吗

不到啊，我感觉也许？可以？

**Comment by boomzero** [comment]
@claude review

**Comment by PythonSmall-Q** [comment]
我觉得问题不大，我看了看不会影响。我个人觉得可以了











在 2025-12-08 20:59:34，"claude[bot]" ***@***.***> 写道：

claude[bot] left a comment (XMOJ-Script-dev/xmoj-bbs-v2#4)
Pull Request Review: Core BBS Routes and Utilities
Summary

This is a substantial rewrite bringing the XMOJ BBS to Nitro/Cloudflare Workers. The implementation demonstrates good security awareness and includes proper input validation, SQL injection protection, and XSS prevention. However, there are several areas that need attention before merging.

Critical Issues
1. SQL Injection Risk via Operator Parameter

Location: server/utils/database.ts:137-191

The Condition parameter accepts an Operator field that is directly concatenated into SQL queries:

QueryString+="`"+i+"` "+Condition[i]["Operator"]+" ? AND ";

Risk: While table/column names are validated, the operator is not. An attacker could potentially inject SQL through malicious operators.

Recommendation:

Whitelist allowed operators: =, !=, <, >, <=, >=, LIKE, IN
Validate operator before concatenation:
constALLOWED_OPERATORS=['=','!=','<','>','<=','>=','LIKE','IN'];if(!ALLOWED_OPERATORS.includes(Condition[i]["Operator"])){thrownewError('Invalid operator');}
2. Session Token Exposure in Logs

Location: server/utils/auth.ts:44-48, 62, 136-137, 166

Session tokens are logged with masking, but the masking function only obscures part of the token. This could still aid in brute-force attacks.

Recommendation:

Avoid logging session tokens entirely, or use cryptographically secure hashing
Remove mask() calls from debug logging
Consider using request IDs for correlation instead
3. Missing Authentication Check

Location: server/middleware/1.auth.ts:39-41

The middleware returns early for non-POST requests without authentication, but some routes may use GET/PUT/DELETE and still require auth.

Recommendation:

Review all routes to ensure they use POST if they require auth
Or modify middleware to handle other HTTP methods that need authentication
4. Hardcoded Admin User List

Location: server/utils/auth.ts:32-35

Admin users, silenced users, and deny lists are hardcoded in the source code:

constAdminUserList: Array<string>=["chenlangning","shanwenxiao","zhuchenrui2"];

Risk:

No audit trail for permission changes
Requires code deployment to update permissions
Exposed in client-side code bundles

Recommendation:

Move to database table with proper audit logging
Add admin panel for managing permissions
At minimum, move to environment variables
⚠️ Security Concerns
5. SSRF Vulnerability in Session Validation

Location: server/utils/auth.ts:108-139

The code fetches from https://www.xmoj.tech to validate sessions. This external call:

Could be exploited if the session system is compromised
Has no timeout configured
Could cause cascading failures if xmoj.tech is down

Recommendation:

Add request timeout (5-10 seconds)
Implement circuit breaker pattern for external calls
Consider caching validation results more aggressively
6. Insufficient Rate Limiting

Location: server/middleware/0.rate-limit.ts:7-8

Current limits: 30 operations with 10/sec refill

May be too permissive for expensive operations (image upload, post creation)
No differentiation between read/write operations
Anonymous users have same limits as authenticated users

Recommendation:

Implement tiered rate limiting (different limits per endpoint)
Stricter limits for unauthenticated users
Consider separate limits for resource-intensive operations
7. Image Upload Validation

Location: server/routes/UploadImage.ts:29-36

Good start, but needs improvement:

constdataUrlRegex=/^data:image\/(png|jpe?g|gif|webp);base64,[A-Za-z0-9+/]+=*$/i

Issues:

Regex validation alone doesn't verify actual image content
Could accept malicious payloads disguised as images
No virus/malware scanning

Recommendation:

Use proper image validation library to verify file headers
Consider using Cloudflare Images or similar service for additional scanning
Implement Content Security Policy headers
8. XSS Protection in GetPost Route

Location: server/routes/GetPost.ts:86-87

Domain replacement happens AFTER content is retrieved:

processedContent=processedContent.replace(/xmoj-bbs\.tech/g,"xmoj-bbs.me");

Issue: This doesn't sanitize output, only does domain replacement. If unsanitized content entered DB before sanitize-html was added, it could still contain XSS payloads.

Recommendation:

Run a data migration to sanitize existing content
Add output sanitization as defense-in-depth
Consider Content Security Policy headers
Code Quality Issues
9. TypeScript Type Safety

Multiple locations use @ts-nocheck or any types:

server/utils/database.ts:1 - entire file has @ts-nocheck
server/middleware/1.auth.ts:26 - event: any
Many utility functions return/accept any

Impact: Defeats the purpose of using TypeScript

Recommendation:

Remove @ts-nocheck and properly type the Database class
Create proper type definitions for Cloudflare bindings
Use unknown instead of any where types are truly dynamic
10. Error Handling Inconsistencies

Location: Various route files

Some routes use:

ThrowErrorIfFailed() (throws on failure)
Direct return new Result(false, ...) (returns failure)
Mix of both patterns

Recommendation:

Standardize on one approach (preferably throwing and catching in error handler)
Document the error handling pattern in CLAUDE.md
11. Missing Input Validation Ranges

Location: server/routes/GetPost.ts:28-29, server/routes/NewPost.ts:30-34

Numeric inputs like PostID and ProblemID have no range validation.

Recommendation:

Add min/max constraints to CheckParams:
CheckParams(Data,{"PostID": {type: "number",min: 1},"Page": {type: "number",min: 1}})
12. Global State Management

Location: server/utils/auth.ts:86, server/middleware/0.rate-limit.ts:37

Uses globalThis for caching:

constglobalCache=(globalThisasany).__tokenCache|| ...

Issues:

Not suitable for multi-isolate Cloudflare Workers
Could cause memory leaks
No eviction policy beyond opportunistic cleanup

Recommendation:

Rely on KV/Durable Objects for distributed state
Document that local caching is dev-only
Or use proper LRU cache library
13. Magic Numbers

Multiple hardcoded values without constants:

Session expiry: 1000 * 60 * 60 * 24 * 7 (server/utils/auth.ts:30)
Pagination: 15 replies per page (server/routes/GetPost.ts:54, 81-82)
Max image size: 5 * 1024 * 1024 (server/routes/UploadImage.ts:46)

Recommendation:

Extract to named constants or configuration file
Strengths
Excellent Security Practices:
✅ SQL Injection Protection: Table/column name whitelisting in database.ts
✅ XSS Prevention: Using sanitize-html for user content
✅ CAPTCHA Integration: Cloudflare Turnstile on post/reply creation
✅ Parameterized Queries: All SQL uses bound parameters
✅ Content Validation: Proper filename and data URL validation
Good Architecture:
✅ Middleware Chain: Clean separation of concerns (rate limit → auth)
✅ Centralized Error Handling: Unified Result pattern
✅ Scheduled Tasks: Automatic cleanup of expired data
✅ Comprehensive CI: Multiple Node versions, TypeScript checks, linting
Documentation:
✅ Excellent README: Clear setup and deployment instructions
✅ OpenAPI Spec: API documentation included
✅ Copyright Headers: Proper licensing (AGPL-3.0)
Test Coverage

Test suite exists but is minimal:

6 test files covering basic functionality
Tests use mocks effectively
Missing integration tests for critical paths

Recommendation:

Add tests for:
SQL injection prevention in database.ts
XSS prevention in sanitizers
Rate limiting behavior
Authentication edge cases (expired tokens, mismatches)
Permission checks (IsAdmin, IsSilenced)
Minor Issues

generate-routes.js references old file:

CheckPrams should be CheckParams (typo was fixed but script not updated)

Unused dependency: sqlstring in package.json appears unused

Missing CORS configuration: No CORS headers configured for cross-origin requests

No request size limits: Missing body size validation could lead to DoS

Recommendations Summary

Before merging:

Fix SQL operator injection vulnerability (Critical)
Remove/reduce session token logging (Critical)
Move admin users to database or env vars (Critical)
Add request timeouts to external calls (High)
Implement proper TypeScript types (High)

Post-merge improvements:

Expand test coverage significantly
Add tiered rate limiting
Implement proper caching strategy for Workers
Add request size limits
Configure CORS properly
Data migration for existing content sanitization
✅ Verdict

Conditional Approval - This is quality work with good security foundations, but the critical issues (especially SQL operator injection and hardcoded permissions) should be addressed before merging to production.

The codebase shows good understanding of security principles and the architecture is clean. With the recommended fixes, this will be production-ready.

Estimated effort to address critical issues: 4-8 hours

Great work on this substantial rewrite!

**Comment by boomzero** [comment]
SQL Injection fix一下吧

**Comment by boomzero** [comment]
或者 @ claude fix

**Comment by boomzero** [comment]
@claude fix Missing Authentication in UploadImage

**Comment by PythonSmall-Q** [comment]
@claude fix SQL Injection Risk in GetPosts.ts

**Comment by PythonSmall-Q** [comment]
@claude fix 1,2,3,4,6

**Comment by boomzero** [comment]
逆天

**Comment by boomzero** [comment]
用了多少token...

**Comment by PythonSmall-Q** [comment]
你要不看看现在行不行，先deploy测试一下

**Comment by boomzero** [comment]
Wait till tomorrow

**Comment by PythonSmall-Q** [comment]
You fix that?

**Comment by PythonSmall-Q** [comment]
@claude fix 1,2,3,4,6,7,8

**Comment by PythonSmall-Q** [comment]
@copilot review

**Comment by PythonSmall-Q** [comment]
@copilot fix all problems in this pull request

**Comment by PythonSmall-Q** [comment]
@copilot open a new pull request to apply changes based on the comments in [this thread](https://github.com/XMOJ-Script-dev/xmoj-bbs-v2/pull/4#pullrequestreview-3574100487) std database looks like this: CREATE TABLE std_answer (
    problem_id INTEGER PRIMARY KEY NOT NULL,
    std_code TEXT
);

**Comment by PythonSmall-Q** [comment]
@langningchen @boomzero

**Comment by PythonSmall-Q** [comment]
@cubic-dev-ai review this

**Comment by boomzero** [comment]
Unfortunately no websocket implementation

**Comment by boomzero** [comment]
What an enormous diff

**Comment by boomzero** [comment]
Anyways, I made some changes to the old backend which should probably be synced here

**Comment by boomzero** [comment]
Woah you used a lot of Claude tokens

**Comment by PythonSmall-Q** [comment]
给我改郁闷了

**Comment by boomzero** [comment]
image
Stop bro, you are using real money now

**Comment by boomzero** [comment]
> image Stop bro, you are using real money now

image

Like, actual real money

**Comment by boomzero** [comment]
image

Oof

**Comment by PythonSmall-Q** [comment]
你先关一下additionalusage

**Comment by boomzero** [comment]
Disabled workflow

**Comment by PythonSmall-Q** [comment]
我觉得差不多了，要不部署测试一下？

**Comment by boomzero** [comment]
Okay, but not to prod pls

**Comment by boomzero** [comment]
How about testing locally?

File: server/routes/GetBadge.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/* Copyright header omitted */
```
Comment (boomzero) [inline server/routes/GetBadge.ts]: ?

Comment (PythonSmall-Q) [reply server/routes/GetBadge.ts]: AI干的

File: server/middleware/0.rate-limit.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  if (st && (now - st.last > TTL_MS)) {
+    globalBuckets.delete(key);
+    st = undefined;
+  }
+  st = st || { tokens: CAPACITY, last: now };
+  const elapsed = (now - st.last) / 1000;
+  st.tokens = Math.min(CAPACITY, st.tokens + elapsed * REFILL_PER_SEC);
+  st.last = now;
+  if (st.tokens < 1) {
+    throw createError({ statusCode: 429, message: '请求过于频繁，请稍后重试' });
```
Comment (boomzero) [inline server/middleware/0.rate-limit.ts]: Do we really need rate limiting? We can configure it at the Cloudflare level, and it would not count towards workers usage.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  build-and-test:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        node: [18, 20, 22]
```
Comment (boomzero) [inline .github/workflows/ci.yml:14]: We only need to support one version, currently 20.
https://github.com/XMOJ-Script-dev/XMOJ-bbs/actions/runs/22307076070/job/64528953025

**Comment by PythonSmall-Q** [comment]
你要不帮帮我秀秀bug叭我没招了

**Comment by boomzero** [comment]
> 你要不帮帮我秀秀bug叭我没招了

Sadly I have a lot of homework to do...

**Comment by PythonSmall-Q** [comment]
@copilot check and fix all bugs discussed in this pr. add new features from XMOJ-Script-dev/XMOJ-bbs that don't exist here. remove rate limiting;
